internal/collector/scraper.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 SAP SE
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-bits/logg"
+)
+
+// How long to wait before re-running the scraper for all tasks.
+const scrapeInterval = 1 * time.Minute
+
+// Scraper holds a collection of Task(s) and other parameters that are required for
+// scraping (read: updating) the metric values.
+type Scraper struct {
+	MaxFailures      int
+	Tasks            map[string]Task                 // key = task name
+	FailureCount     map[string]int                  // map of task name to its failure count
+	ExitCodeGaugeVec map[string]*prometheus.GaugeVec // map of task name to its relevant exit code GaugeVec
+}
+
+// NewScraper returns a new Scraper.
+func NewScraper(maxFailures int) *Scraper {
+	return &Scraper{
+		MaxFailures:      maxFailures,
+		Tasks:            make(map[string]Task),
+		FailureCount:     make(map[string]int),
+		ExitCodeGaugeVec: make(map[string]*prometheus.GaugeVec),
+	}
+}
+
+// Run updates the metrics for all tasks periodically as per the scrapeInterval.
+func (s *Scraper) Run() {
+	for {
+		s.UpdateAllMetrics()
+		time.Sleep(scrapeInterval)
```

Comment:
`UpdateAllMetrics` may take quite a while, so we should factor its runtime into the scrape cycle.
```suggestion
		startedAt := time.Now()
		s.UpdateAllMetrics()
		//slow down if UpdateAllMetrics() finished faster than the desired scrape interval
		sleepDuration := scrapeInterval.Sub(time.Since(startedAt))
		if sleepDuration > 0 {
			time.Sleep(sleepDuration)
		}
```

Reply:
I messed up the variable naming here. `scrapeInterval` which has a value of `1 * time.Minute` should be called `idleInterval`.

Then the following:

```
s.UpdateAllMetrics()
time.Sleep(idleInterval)
```

would mean an idle time of 1 minute between each scrape cycle, regardless of how long the scrape takes.

Reply:
That's true, but I'd still rather want the behavior I described. If we do 10 different subtasks, I don't want the time between subtask 10 and the next subtask 1 to be one minute. I want the time between subtask 1 and the next subtask 1 to be one minute. Basing the sleep off the _start_ of the previous scrape better approximates that target behavior.